prusti-viper/src/encoder/mir/pure/specifications/encoder.rs

```diff
@@ -255,7 +255,7 @@ impl<'p, 'v: 'p, 'tcx: 'v> SpecificationEncoder<'p, 'v, 'tcx> {
                                         self.encode_quantifier_arg(
                                             *arg,
                                             arg_ty,
-                                            &format!("{}_{}", vars.spec_id, vars.pre_id),
```

Comment:
Have you checked whether this does not lead to strange name conflicts in encoded programs that are insanely hard to understand on the source level?

Reply:
It would be great to have a bunch of tests that check this.

Reply:
I think that might be possible; I'll take a look at creating some test cases for this. However, in general I think this is very hard to avoid with deterministic names - one could just look at the generated Viper code and name another local variable in a way which clashes. The two solutions to avoid this that I see: (A) use hash of the HIR/MIR of the function as a sort of 'spec_id' which could be used in these cases (then if another variable is defined after the fact, this hash will change) or (B) look at all defined variables in scope and pick a name that doesn't clash.

Reply:
> (B) look at all defined variables in scope and pick a name that doesn't clash.

We could do this when lowering to `vir_legacy` – at that point, we have all variable names present.

Reply:
> one could just look at the generated Viper code and name another local variable in a way which clashes

Would that still be possible if we add a prefix `quantified_` (or just `q_`) to all quantified variables?

Even with that , as Vytautas suggested I wonder what happens to the encoding of `forall x: u32 :: forall x: i32 :: ...`. Will the Viper names clash?

Reply:
Aha I forgot that the user doesn't have control over MIR names. In that case we probably don't need to resort to the drastic measures I mention above? I'll think about where this might not hold

Reply:
FYI in the spec refactor I base this ID on the `DefId` of the closure it comes from.

Reply:
@Aurel300 do you mean the `spec_id`? The `pre_id` is just the level of nesting and shouldn't change in the refactor, right?

Reply:
@Aurel300 Hmm, but using the `DefId` is problematic in this case since it won't be stable if other stuff in the file/crate is added/deleted. I'm not exactly sure what position in quantifiers args (`arg_idx`) means (do you mean that for a `forall(|a: i32, b: i32| forall(|c: i32| ...))`, then `a` is 1, `b` is 2, `c` is 1?), but could we not use only this, with also a `quantifier_nesting_count` in the name as well?

Reply:
Yes, it wasn't chosen with caching in mind, and it is still better than the original approach of random UUIDs. A nesting count would work, but I'm not sure it would be easy to introduce? When encoding specifications into Viper expressions we would generally work inside out, so it's hard for the inner `forall` to know its nesting depth.

Reply:
The current `vars.pre_id` seems to do this now though (at least from looking at the generated Viper), I'm not sure how it's implemented but maybe if we just copied the ideas from that?

Reply:
I though `vars.pre_id` corresponded to another specification UUID? Maybe it's being interned somewhere?